File: components/server/src/ide-config.ts
/**
 * Copyright (c) 2021 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License-AGPL.txt in the project root for license information.
 */

import { Disposable, DisposableCollection, Emitter } from '@gitpod/gitpod-protocol';
import { filePathTelepresenceAware } from '@gitpod/gitpod-protocol/lib/env';
import { IDEClient, IDEOptions } from '@gitpod/gitpod-protocol/lib/ide-protocol';
import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
import { repeat } from '@gitpod/gitpod-protocol/lib/util/repeat';
import * as Ajv from 'ajv';
import * as cp from 'child_process';
import * as crypto from 'crypto';
import * as fs from 'fs';
import { injectable } from 'inversify';
import debounce = require('lodash.debounce');

export interface IDEConfig {
    supervisorImage: string;
    ideOptions: IDEOptions;
    clients?: { [id: string]: IDEClient };
}

const scheme = {
    type: 'object',
    properties: {
        supervisorImage: {
            type: 'string',
        },
        ideOptions: {
            type: 'object',
            properties: {
                options: {
                    type: 'object',
                    additionalProperties: {
                        type: 'object',
                        properties: {
                            orderKey: { type: 'string' },
                            title: { type: 'string' },
                            type: { type: 'string' },
                            logo: { type: 'string' },
                            tooltip: { type: 'string' },
                            label: { type: 'string' },
                            nodes: { type: 'array', items: { type: 'string' } },
                            hidden: { type: 'boolean' },
                            image: { type: 'string' },
                            latestImage: { type: 'string' },
                            resolveImageDigest: { type: 'boolean' },
                        },
                        required: ['title', 'type', 'logo', 'image'],
                    },
                },
                defaultIde: { type: 'string' },
                defaultDesktopIde: { type: 'string' },
                clients: {
                    type: 'object',
                    additionalProperties: {
                        type: 'object',
                        properties: {
                            defaultDesktopIDE: { type: 'string' },
                            desktopIDEs: { type: 'array', items: { type: 'string' } },
                        },
                    },
                },
            },
            required: ['options', 'defaultIde', 'defaultDesktopIde'],
        },
    },
    required: ['supervisorImage', 'ideOptions'],
};

@injectable()
export class IDEConfigService {
    private readonly ajv = new Ajv();
    private readonly configPath: string;
    private readonly validate: Ajv.ValidateFunction;

    private state: {
        key?: string;
        value?: IDEConfig;
    } = {};
    private readonly onDidChangeEmitter = new Emitter<IDEConfig>();

    constructor() {
        const configPath = process.env.IDE_CONFIG_PATH;
        if (!configPath) {
            throw new Error('ide config: IDE_CONFIG_PATH not set');
        }
        this.configPath = filePathTelepresenceAware(configPath);
        this.validate = this.ajv.compile(scheme);
        this.reconcile('initial');
        fs.watchFile(this.configPath, (curr, prev) => {
            if (curr.mtimeMs != prev.mtimeMs) {
                this.reconcile('file changed');
            }
        });
        repeat(() => this.reconcile('interval'), 60 * 60 * 1000 /* 1 hour */);
    }

    get config(): Promise<IDEConfig> {
        if (this.state.value) {
            return Promise.resolve(this.state.value);
        }
        const toFinalize = new DisposableCollection();
        return new Promise<IDEConfig>((resolve, reject) => {
            const timeoutHandle = setTimeout(() => reject(new Error('ide config: read timeout')), 30 * 1000);
            toFinalize.push(Disposable.create(() => clearTimeout(timeoutHandle)));
            toFinalize.push(this.onDidChangeEmitter.event(resolve));
        }).finally(() => toFinalize.dispose());
    }

    private contentHash: string | undefined;
    private reconcile = debounce(
        async (trigger: string) => {
            try {
                let fileContent: string | undefined;
                try {
                    fileContent = await fs.promises.readFile(this.configPath, { encoding: 'utf-8' });
                } catch {}
                if (!fileContent) {
                    return;
                }

                let value = this.state.value;

                const newValue: IDEConfig = JSON.parse(fileContent);
                const contentHash = crypto.createHash('sha256').update(fileContent, 'utf8').digest('hex');
                const contentChanged = this.contentHash !== contentHash;
                if (contentChanged) {
                    this.contentHash = contentHash;

                    this.validate(newValue);
                    if (this.validate.errors) {
                        throw new Error('invalid: ' + this.ajv.errorsText(this.validate.errors));
                    }

                    if (!(newValue.ideOptions.defaultIde in newValue.ideOptions.options)) {
                        throw new Error(
                            `invalid: There is no IDEOption entry for editor '${newValue.ideOptions.defaultIde}'.`,
                        );
                    }
                    if (!(newValue.ideOptions.defaultDesktopIde in newValue.ideOptions.options)) {
                        throw new Error(
                            `invalid: There is no IDEOption entry for default desktop IDE '${newValue.ideOptions.defaultDesktopIde}'.`,
                        );
                    }
                    if (newValue.ideOptions.options[newValue.ideOptions.defaultIde].type != 'browser') {
                        throw new Error(
                            `invalid: Editor '${
                                newValue.ideOptions.defaultIde
                            }' needs to be of type 'browser' but is '${
                                newValue.ideOptions.options[newValue.ideOptions.defaultIde].type
                            }'.`,
                        );
                    }
                    if (newValue.ideOptions.options[newValue.ideOptions.defaultDesktopIde].type != 'desktop') {
                        throw new Error(
                            `invalid: Editor (desktop), '${
                                newValue.ideOptions.defaultDesktopIde
                            }' needs to be of type 'desktop' but is '${
                                newValue.ideOptions.options[newValue.ideOptions.defaultIde].type
                            }'.`,
                        );
                    }

                    if (newValue.ideOptions.clients) {
                        for (const [clientId, client] of Object.entries(newValue.ideOptions.clients)) {
                            if (
                                client.defaultDesktopIDE &&
                                !(client.defaultDesktopIDE in newValue.ideOptions.options)
                            ) {
                                throw new Error(
                                    `${clientId} client: there is no option entry for editor '${client.defaultDesktopIDE}'.`,
                                );
                            }
                            if (client.desktopIDEs) {
                                for (const ide of client.desktopIDEs) {
                                    if (!(ide in newValue.ideOptions.options)) {
                                        throw new Error(
                                            `${clientId} client: there is no option entry for editor '${ide}'.`,
                                        );
                                    }
                                }
                            }
                        }
                    }

                    value = newValue;
                }

                if (!value) {
                    return;
                }

                // we only want to resolve image by interval or content changed
                if (!(contentChanged || trigger === 'interval')) {
                    return;
                }

                for (const [id, option] of Object.entries(newValue.ideOptions.options).filter(
                    ([_, x]) => !!x.resolveImageDigest,
                )) {
                    try {
                        value.ideOptions.options[id].image = await this.resolveImageDigest(option.image);
                        log.info('ide config: successfully resolved image digest', {
                            ide: id,
                            image: option.image,
                            resolvedImage: value.ideOptions.options[id].image,
                            trigger,
                        });
                    } catch (e) {
                        log.error('ide config: error while resolving image digest', e, { trigger });
                    }
                }

                for (const [id, option] of Object.entries(newValue.ideOptions.options).filter(
                    ([_, x]) => x.latestImage,
                )) {
                    try {
                        value.ideOptions.options[id].latestImage = await this.resolveImageDigest(option.latestImage!);
                        log.info('ide config: successfully resolved latest image digest', {
                            ide: id,
                            latestImage: option.latestImage,
                            resolvedImage: value.ideOptions.options[id].latestImage,
                            trigger,
                        });
                    } catch (e) {
                        log.error('ide config: error while resolving latest image digest', e, { trigger });
                    }
                }

                const key = JSON.stringify(value);
                if (key === this.state.key) {
                    return;
                }

                log.info('ide config: updated', { newConfig: JSON.stringify(value, undefined, 2), trigger });
                this.state = { key, value };
                this.onDidChangeEmitter.fire(value);
            } catch (e) {
                log.error('ide config: failed to reconcile', e, { trigger });
            }
        },
        500,
        { leading: true },
    );

    private resolveImageDigest(imageName: string) {
        return new Promise<string>((resolve, reject) => {
            cp.exec(`oci-tool --timeout 30s resolve name ${imageName}`, (error, imageDigest) => {
                if (error) {
                    return reject(error);
                }

                if (!imageDigest) {
                    throw new Error(`Cannot resolve ${imageName} image`);
                }

                return resolve(imageDigest.trim());
            });
        });
    }
}

File: components/ide-service/pkg/server/ideconfig.go
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License.AGPL.txt in the project root for license information.

package server

import (
	"context"
	_ "embed"
	"encoding/json"
	"fmt"

	"github.com/containerd/containerd/remotes"
	"github.com/gitpod-io/gitpod/common-go/log"
	"github.com/gitpod-io/gitpod/ide-service-api/config"
	oci_tool "github.com/gitpod-io/gitpod/ide-service/pkg/ocitool"
	"github.com/xeipuuv/gojsonschema"
	"golang.org/x/xerrors"
)

//go:embed schema.json
var jsonScheme []byte

// ParseConfig parse and validate ide config
func ParseConfig(ctx context.Context, res remotes.Resolver, b []byte) (*config.IDEConfig, error) {
	var cfg config.IDEConfig
	if err := json.Unmarshal(b, &cfg); err != nil {
		return nil, xerrors.Errorf("cannot parse ide config: %w", err)
	}

	// validate with json schema
	schemaLoader := gojsonschema.NewBytesLoader(jsonScheme)
	documentLoader := gojsonschema.NewBytesLoader(b)

	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
	if err != nil {
		return nil, xerrors.Errorf("invalid: %w", err)
	}

	if !result.Valid() {
		log.Error("invalid ide config")
		for _, desc := range result.Errors() {
			log.WithField("desc", desc).Error("invalid ide config desc")
		}
		return nil, xerrors.New("invalid ide config")
	}

	if err := checkIDEExistsInOptions(cfg, cfg.IdeOptions.DefaultIde, config.IDETypeBrowser, "DefaultIde"); err != nil {
		return nil, err
	}
	if err := checkIDEExistsInOptions(cfg, cfg.IdeOptions.DefaultDesktopIde, config.IDETypeDesktop, "DefaultDesktopIde"); err != nil {
		return nil, err
	}

	for clientId, client := range cfg.IdeOptions.Clients {
		if err := checkIDEExistsInOptions(cfg, client.DefaultDesktopIDE, config.IDETypeDesktop, fmt.Sprintf("client %s DefaultDesktopIDE", clientId)); err != nil {
			return nil, err
		}
		for _, ideId := range client.DesktopIDEs {
			if err := checkIDEExistsInOptions(cfg, ideId, config.IDETypeDesktop, fmt.Sprintf("client %s DesktopIDEs %s", clientId, ideId)); err != nil {
				return nil, err
			}
		}
	}

	// resolve image digest
	for id, option := range cfg.IdeOptions.Options {
		if option.ResolveImageDigest {
			if resolved, err := oci_tool.Resolve(ctx, res, option.Image); err != nil {
				log.WithError(err).Error("ide config: cannot resolve image digest")
			} else {
				log.WithField("ide", id).WithField("image", option.Image).WithField("resolved", resolved).Info("ide config: resolved latest image digest")
				option.Image = resolved
			}
		}
		if resolvedVersion, err := oci_tool.ResolveIDEVersion(ctx, res, option.Image); err != nil {
			log.WithError(err).Error("ide config: cannot get version from image")
		} else {
			option.ImageVersion = resolvedVersion
		}
		if option.LatestImage != "" {
			if resolved, err := oci_tool.Resolve(ctx, res, option.LatestImage); err != nil {
				log.WithError(err).Error("ide config: cannot resolve latest image digest")
			} else {
				log.WithField("ide", id).WithField("image", option.LatestImage).WithField("resolved", resolved).Info("ide config: resolved latest image digest")
				option.LatestImage = resolved
			}
			if resolvedVersion, err := oci_tool.ResolveIDEVersion(ctx, res, option.LatestImage); err != nil {
				log.WithError(err).Error("ide config: cannot get version from image")
			} else {
				option.LatestImageVersion = resolvedVersion
			}
		}

		// append or replace latest stable version into versions
		if option.AllowPin && option.ImageVersion != "" {
			found := false
			foundIndex := -1
			for index, version := range option.Versions {
				if version.Version == option.ImageVersion {
					found = true
					foundIndex = index
					break
				}
			}
			currentVersion := config.IDEVersion{
				Version:     option.ImageVersion,
				Image:       option.Image,
				ImageLayers: option.ImageLayers,
			}
			if !found {
				var versions []config.IDEVersion
				versions = append(versions, currentVersion)
				versions = append(versions, option.Versions...)
				option.Versions = versions
			} else if foundIndex >= 0 {
				versions := append([]config.IDEVersion{}, option.Versions...)
				versions[foundIndex] = currentVersion
				option.Versions = versions
			}
		}
		cfg.IdeOptions.Options[id] = option
	}

	return &cfg, nil
}

func checkIDEExistsInOptions(c config.IDEConfig, ideId string, ideType config.IDEType, name string) error {
	ide, ok := c.IdeOptions.Options[ideId]
	if !ok {
		return xerrors.Errorf("invalid ide config: %s %s is not an entry of ide options", name, ideId)
	}
	if ide.Type != ideType {
		return xerrors.Errorf("invalid ide config: %s should be %s but %s", name, ideType, ide.Type)
	}
	return nil
}
